File: Reference_Implementation/lib/LESS.c
/**
 *
 * Reference ISO-C11 Implementation of LESS.
 *
 * @version 1.2 (February 2025)
 *
 * @author Alessandro Barenghi <alessandro.barenghi@polimi.it>
 * @author Gerardo Pelosi <gerardo.pelosi@polimi.it>
 * @author Floyd Zweydinger <zweydfg8+github@rub.de>
 *
 * This code is hereby placed in the public domain.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHORS ''AS IS'' AND ANY EXPRESS
 * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 **/
#include <string.h> // memcpy, memset
#include "LESS.h"
#include "canonical.h"
#include "seedtree.h"
#include "rng.h"
#include "utils.h"
#include "fips202.h"
#include "sha3.h"

void LESS_keygen(prikey_t *SK,
                 pubkey_t *PK) {
    /* generating private key from a single seed */
    randombytes(SK->compressed_sk, PRIVATE_KEY_SEED_LENGTH_BYTES);

    /* expanding it onto private seeds */
    SHAKE_STATE_STRUCT sk_shake_state;
    initialize_csprng(&sk_shake_state, SK->compressed_sk, PRIVATE_KEY_SEED_LENGTH_BYTES);

    /* Generating public code G_0 */
    csprng_randombytes(PK->G_0_seed, SEED_LENGTH_BYTES, &sk_shake_state);

    rref_generator_mat_t G0_rref;
    generator_sample(&G0_rref, PK->G_0_seed);

    generator_mat_t tmp_full_G;
    generator_rref_expand(&tmp_full_G, &G0_rref);

    /* The first private key monomial is an ID matrix, no need for random
     * generation, hence NUM_KEYPAIRS-1 */
    unsigned char private_monomial_seeds[NUM_KEYPAIRS - 1][PRIVATE_KEY_SEED_LENGTH_BYTES];
    for (uint32_t i = 0; i < NUM_KEYPAIRS - 1; i++) {
        csprng_randombytes(private_monomial_seeds[i],
                           PRIVATE_KEY_SEED_LENGTH_BYTES,
                           &sk_shake_state);
    }

    /* note that the first "keypair" is just the public generator G_0, stored
     * as a seed and the identity matrix (not stored) */
    for (uint32_t i = 0; i < NUM_KEYPAIRS - 1; i++) {
        uint8_t is_pivot_column[N_pad] = {0};
        /* expand inverse monomial from seed */
        monomial_t private_Q;
        monomial_t private_Q_inv;
        monomial_sample_prikey(&private_Q_inv, private_monomial_seeds[i]);
        monomial_inv(&private_Q, &private_Q_inv);

        generator_mat_t result_G = {0};
        generator_monomial_mul(&result_G,
                               &tmp_full_G,
                               &private_Q);
        memset(is_pivot_column, 0, sizeof(is_pivot_column));
        generator_RREF(&result_G, is_pivot_column);
        /* note that the result is stored at i-1 as the first
         * public key element is just a seed */
        compress_rref(PK->SF_G[i],
                      &result_G,
                      is_pivot_column);
    }
} /* end LESS_keygen */

/// returns the number of opened seeds in the tree.
/// \param SK[in]: secret key
/// \param m[in]: message to sign
/// \param mlen[in]: length of the message to sign in bytes
/// \param sig[out]: signature
/// \return: x: number of leaves opened by the algorithm
size_t LESS_sign(const prikey_t *SK,
                 const char *const m,
                 const uint64_t mlen,
                 sign_t *sig) {
    uint8_t g0_initial_pivot_flags [N_pad];
    uint8_t is_pivot_column[N_pad];

    /*         Private key expansion        */
    /* expand sequence of seeds for private inverse-monomial matrices */
    SHAKE_STATE_STRUCT sk_shake_state;
    initialize_csprng(&sk_shake_state, SK->compressed_sk, PRIVATE_KEY_SEED_LENGTH_BYTES);

    /* Generating seed for public code G_0 */
    unsigned char G_0_seed[SEED_LENGTH_BYTES];
    csprng_randombytes(G_0_seed, SEED_LENGTH_BYTES, &sk_shake_state);

    /* The first private key monomial is an ID matrix, no need for random
     * generation, hence NUM_KEYPAIRS-1 */
    unsigned char private_monomial_seeds[NUM_KEYPAIRS - 1][PRIVATE_KEY_SEED_LENGTH_BYTES];
    for (uint32_t i = 0; i < NUM_KEYPAIRS - 1; i++) {
        csprng_randombytes(private_monomial_seeds[i],
                           PRIVATE_KEY_SEED_LENGTH_BYTES,
                           &sk_shake_state);
    }

    // generate the salt from a TRNG
    randombytes(sig->salt, HASH_DIGEST_LENGTH);

    /*         Ephemeral monomial generation        */
    unsigned char ephem_monomials_seed[SEED_LENGTH_BYTES];
    csprng_randombytes(ephem_monomials_seed,
                       SEED_LENGTH_BYTES,
                       &sk_shake_state);

    /* create the prng for the "blinding" monomials for the canonical form computation */
    uint8_t cf_seed[SEED_LENGTH_BYTES];
    csprng_randombytes(cf_seed,
                       SEED_LENGTH_BYTES,
                       &sk_shake_state);
    SHAKE_STATE_STRUCT cf_shake_state;
    initialize_csprng(&cf_shake_state, cf_seed, SEED_LENGTH_BYTES);

    unsigned char seed_tree[NUM_NODES_SEED_TREE * SEED_LENGTH_BYTES] = {0};
    BuildGGM(seed_tree, ephem_monomials_seed, sig->salt);

    unsigned char linearized_rounds_seeds[T*SEED_LENGTH_BYTES] = {0};
    seed_leaves(linearized_rounds_seeds,seed_tree);

    /*         Public G_0 expansion                  */
    rref_generator_mat_t G0_rref;
    generator_sample(&G0_rref, G_0_seed);
    generator_get_pivot_flags (&G0_rref, g0_initial_pivot_flags);
    generator_mat_t full_G0, G0;
    generator_rref_expand(&full_G0, &G0_rref);

    monomial_t mu_tilde;
    monomial_action_IS_t pi_tilde[T];
    normalized_IS_t A_i = {0};

    LESS_SHA3_INC_CTX state;
    LESS_SHA3_INC_INIT(&state);

    for (uint32_t i = 0; i < T; i++) {
        monomial_sample_salt(&mu_tilde,
                             linearized_rounds_seeds + i * SEED_LENGTH_BYTES,
                             sig->salt,
                             i);
        generator_monomial_mul(&G0, &full_G0, &mu_tilde);
        memset(is_pivot_column, 0, N_pad);
#if defined(LESS_REUSE_PIVOTS_SG)
        uint8_t permuted_pivot_flags[N_pad];
        for (uint32_t t = 0; t < N; t++) {
            permuted_pivot_flags[mu_tilde.permutation[t]] = g0_initial_pivot_flags[t];
        }
        if (generator_RREF_pivot_reuse(&G0, is_pivot_column, permuted_pivot_flags, SIGN_PIVOT_REUSE_LIMIT) == 0) {
            return 0;
        }
#else
        if (generator_RREF(&G0, is_pivot_column) == 0) {
            return 0;
        }
#endif
        // just copy the non IS
        uint32_t ctr = 0;
        for(uint32_t j = 0; j < N-K; j++) {
            while (is_pivot_column[ctr]) {
                ctr += 1;
            }
            /// copy column
            for (uint32_t k = 0; k < K; k++) {
                A_i.values[k][j] = G0.values[k][ctr];
            }
            ctr += 1;
        }

        POSITION_T piv_idx = 0;
        for(uint32_t col_idx = 0; col_idx < N; col_idx++) {
            POSITION_T row_idx = 0;
            for(uint32_t t = 0; t < N; t++) {
               if (mu_tilde.permutation[t] == col_idx) {
                   row_idx = t;
                   break;
               }
            }
            if(is_pivot_column[col_idx] == 1) {
               pi_tilde[i].permutation[piv_idx] = row_idx;
               piv_idx++;
            }
        }

        blind(&A_i, &cf_shake_state);
        const int t = CF(&A_i);

        if (t == 0) {
            *(linearized_rounds_seeds + i*SEED_LENGTH_BYTES) += 1;
            i -= 1;
        } else {
            // NOTE: as we increase the size of the `normalized_IS_t`
            // we need to hash the values row by row.
#if defined(USE_AVX2) || defined(USE_NEON)
            for (uint32_t sl = 0; sl < K; sl++) {
                LESS_SHA3_INC_ABSORB(&state, A_i.values[sl], K);
            }
#else
            LESS_SHA3_INC_ABSORB(&state, (uint8_t *)&A_i, sizeof(normalized_IS_t));
#endif
        }
    }

    LESS_SHA3_INC_ABSORB(&state, (const uint8_t *)m, mlen);
    LESS_SHA3_INC_ABSORB(&state, sig->salt, HASH_DIGEST_LENGTH);

    /* Squeeze output */
    LESS_SHA3_INC_FINALIZE(sig->digest, &state);
    // (b_0, ..., b_{t-1})
    uint8_t fixed_weight_string[T] = {0};
    SampleChallenge(fixed_weight_string, sig->digest);

    uint8_t indices_to_publish[T];
    for (uint32_t i = 0; i < T; i++) {
        indices_to_publish[i] = !!(fixed_weight_string[i]);
    }

    int emitted_monoms = 0;
    memset(&sig->seed_storage, 0, SEED_TREE_MAX_PUBLISHED_BYTES);

    const uint32_t num_seeds_published =
            GGMPath(seed_tree,
                           indices_to_publish,
                           (unsigned char *) &sig->seed_storage);

    monomial_action_IS_t mono_action;
    for (uint32_t i = 0; i < T; i++) {
        monomial_t Q_to_multiply;
        if (fixed_weight_string[i] != 0) {
            const int sk_monom_seed_to_expand_idx = fixed_weight_string[i];

            monomial_sample_prikey(&Q_to_multiply,
                                   private_monomial_seeds[sk_monom_seed_to_expand_idx - 1]);
            monomial_compose_action(&mono_action, &Q_to_multiply, &pi_tilde[i]);

            CosetRep(sig->cf_monom_actions[emitted_monoms], &mono_action);
            emitted_monoms++;
        }
    }
    return num_seeds_published;
} /* end LESS_sign */

/// NOTE: non-constant time
/// \param PK[in]: public key
/// \param m[in]: message for which a signature was computed
/// \param mlen[in]: length of the message in bytes
/// \param sig[in]: signature
/// \return 0: on failure
///         1: on success
int LESS_verify(const pubkey_t *const PK,
                const char *const m,
                const uint64_t mlen,
                const sign_t *const sig) {
    uint8_t fixed_weight_string[T] = {0};
    uint8_t is_pivot_column[N_pad];
    uint8_t g0_initial_pivot_flags[N];
    uint8_t gi_initial_pivot_flags[N];
#ifdef LESS_REUSE_PIVOTS_VY
    uint8_t g0_permuted_pivot_flags[N];
#endif
    SampleChallenge(fixed_weight_string, sig->digest);

    uint8_t published_seed_indexes[T];
    for (uint32_t i = 0; i < T; i++) {
        published_seed_indexes[i] = !!(fixed_weight_string[i]);
    }

    unsigned char seed_tree[NUM_NODES_SEED_TREE * SEED_LENGTH_BYTES] = {0};
    uint32_t rebuilding_seeds_went_fine;
    rebuilding_seeds_went_fine = RebuildGGM(seed_tree,
                                                          published_seed_indexes,
                                                          (unsigned char *) &sig->seed_storage,
                                                          sig->salt);
    if (!rebuilding_seeds_went_fine) {
        return 0;
    }

    unsigned char linearized_rounds_seeds[T*SEED_LENGTH_BYTES] = {0};
    seed_leaves(linearized_rounds_seeds,seed_tree);

    int employed_monoms = 0;

    rref_generator_mat_t G0_rref;
    generator_sample(&G0_rref, PK->G_0_seed);

    generator_mat_t G0 = {0}, G0_full = {0};
    generator_mat_t G_prime = {0};
    monomial_t mu_tilde;
    normalized_IS_t Ai = {0};
    LESS_SHA3_INC_CTX state;
    LESS_SHA3_INC_INIT(&state);

    generator_get_pivot_flags(&G0_rref, g0_initial_pivot_flags);
    generator_rref_expand(&G0_full, &G0_rref);

    for (uint32_t i = 0; i < T; i++) {
        memset(is_pivot_column, 0, N_pad);
        if (fixed_weight_string[i] == 0) {
            monomial_sample_salt(&mu_tilde,
                                 linearized_rounds_seeds + i * SEED_LENGTH_BYTES,
                                 sig->salt,
                                 i);

            generator_monomial_mul(&G_prime, &G0_full, &mu_tilde);
#if defined(LESS_REUSE_PIVOTS_VY)
            uint8_t permuted_pivot_flags[N_pad] = {0};

            for (uint32_t t = 0; t < N; t++) {
                permuted_pivot_flags[mu_tilde.permutation[t]] = g0_initial_pivot_flags[t];
            }
            if (generator_RREF_pivot_reuse(&G_prime,is_pivot_column, permuted_pivot_flags, VERIFY_PIVOT_REUSE_LIMIT) == 0) {
                return 0;
            }
#else
            if (generator_RREF(&G_prime, is_pivot_column) == 0) {
                return 0;
            }
#endif
        } else {
            expand_to_rref(&G0, PK->SF_G[fixed_weight_string[i] - 1], gi_initial_pivot_flags);
            if (!CheckCanonicalAction(sig->cf_monom_actions[employed_monoms])) {
                return 0;
            }

#if defined(LESS_REUSE_PIVOTS_VY)
            apply_cf_action_to_G_with_pivots(&G_prime,
                                             &G0,
                                             sig->cf_monom_actions[employed_monoms],
                                             gi_initial_pivot_flags,
                                             g0_permuted_pivot_flags);
            const int ret = generator_RREF_pivot_reuse(&G_prime, is_pivot_column,
                                                       g0_permuted_pivot_flags,
                                                       VERIFY_PIVOT_REUSE_LIMIT);
#else
            apply_cf_action_to_G(&G_prime, &G0, sig->cf_monom_actions[employed_monoms]);
            const int ret = generator_RREF(&G_prime, is_pivot_column);
#endif
            if(ret == 0) {
                return 0;
            }

            employed_monoms++;
        }
        // just copy the non IS
        uint32_t ctr = 0;
        for(uint32_t j = 0; j < N-K; j++) {
            while (is_pivot_column[ctr]) {
                ctr += 1;
            }
            /// copy column
            for (uint32_t k = 0; k < K; k++) {
                Ai.values[k][j] = G_prime.values[k][ctr];
            }
            ctr += 1;
        }
        const int r = CF(&Ai);
        if (r == 0) {
            return 0;
        }
#if defined(USE_AVX2) || defined(USE_NEON)
        for (uint32_t sl = 0; sl < K; sl++) {
            LESS_SHA3_INC_ABSORB(&state, Ai.values[sl], K);
        }
#else
        LESS_SHA3_INC_ABSORB(&state, (uint8_t *)&Ai, sizeof(normalized_IS_t));
#endif
    }

    uint8_t recomputed_digest[HASH_DIGEST_LENGTH] = {0};
    LESS_SHA3_INC_ABSORB(&state, (const uint8_t *) m, mlen);
    LESS_SHA3_INC_ABSORB(&state, sig->salt, HASH_DIGEST_LENGTH);
    /* Squeeze output */
    LESS_SHA3_INC_FINALIZE(recomputed_digest, &state);

    return (verify(recomputed_digest, sig->digest,
                   HASH_DIGEST_LENGTH) == 0);
} /* end LESS_verify */

File: Optimized_Implementation/rust/src/less.rs
#![allow(dead_code)]


/// TODO:
/// type alias: MatrixRREF = Matrix<K, N_pad>
/// type alias: XOF = Shake128

use sha3::{
    Digest, Shake128, Sha3_256,
};
use crate::config::{N, K, T, N_pad, K_pad, N_K_pad, W, N8, PRIVATE_KEY_SEED_LENGTH_BYTES, RREF_MAT_PACKEDBYTES, NUM_KEYPAIRS, HASH_DIGEST_LENGTH, SEED_TREE_MAX_PUBLISHED_BYTES, SEED_LENGTH_BYTES, NUM_NODES_SEED_TREE, SIGN_PIVOT_REUSE_LIMIT};
use crate::matrix::{Matrix, MatrixNormalized};
use crate::monomial::{Monomial, Permutation};
use crate::prng::{
    randombytes,
    csprng_randombytes, csprng_initialize
};
use crate::tree::{build_ggm, ggm_path, seed_leaves};
use crate::utils::sample_challenge;

pub struct PrivateKey {
    compressed_sk: [u8; PRIVATE_KEY_SEED_LENGTH_BYTES]
}

/// Public key: the first gen. matrix is shrunk to just a seed, all the
 /// others are stored in RREF form
pub struct PublicKey {
    g0_seed: [u8; PRIVATE_KEY_SEED_LENGTH_BYTES],
    sf_g: [[u8; RREF_MAT_PACKEDBYTES]; NUM_KEYPAIRS - 1],
}

pub struct Signature {
    digest: [u8; HASH_DIGEST_LENGTH],
    salt: [u8; HASH_DIGEST_LENGTH],
    cf_monom_actions: [[u8; N8]; W],
    seed_storage: [u8; SEED_TREE_MAX_PUBLISHED_BYTES],
}


// LESS_keygen
// \param SK[out]: pointer to an uninitialized secret key data structure
// \param PK[out]: pointer to an uninitialized public key data structure
pub fn less_keygen(sk: &mut PrivateKey, pk: &mut PublicKey) {
    /* generating private key from a single seed */
    randombytes(&mut sk.compressed_sk);

    /* expanding it onto private seeds */
    let mut sk_shake_state = csprng_initialize::<Shake128>(&sk.compressed_sk);

    /* Generating public code G_0 */
    csprng_randombytes(&mut pk.g0_seed, &mut sk_shake_state);

    let g0_rref = Matrix::<K, N_K_pad>::rand_from_seed::<Shake128>(&pk.g0_seed);
    let g0_column_pos: [u16; N_K_pad] = std::array::from_fn(|i| (i + K) as u16);

    let tmp_full_g = Matrix::<K, N_pad>::from_compressed_rref(&g0_column_pos, &g0_rref);

    /* The first private key monomial is an ID matrix, no need for random
     * generation, hence NUM_KEYPAIRS-1 */
    let mut private_monomial_seeds: [[u8; PRIVATE_KEY_SEED_LENGTH_BYTES]; NUM_KEYPAIRS - 1] =
        [[0u8; PRIVATE_KEY_SEED_LENGTH_BYTES]; NUM_KEYPAIRS - 1];
    for i in 0..(NUM_KEYPAIRS - 1) {
        csprng_randombytes(&mut private_monomial_seeds[i], &mut sk_shake_state);
    }
    let mut was_pivot_column = [0u8; N_pad];
    /* note that the first "keypair" is just the public generator G_0, stored
     * as a seed and the identity matrix (not stored) */
    for i in 0..(NUM_KEYPAIRS - 1) {
        let mut is_pivot_column = [0u8; N_pad];
        /* expand inverse monomial from seed */
        let mut private_q = Monomial::<N>::default();
        let private_q_inv = Monomial::<N>::rand_from_seed::<Shake128>(&private_monomial_seeds[i]);
        // NOTE: the inversion is not implemented in constant time.
        Monomial::<N>::inv_non_ct(&mut private_q, &private_q_inv);

        let mut result_g = Matrix::<K, N_pad>::default();
        Matrix::monomial_mul(&mut result_g, &tmp_full_g, &private_q); // TODO replace with `*` operator

        result_g.rref::<false>(&mut is_pivot_column, &mut was_pivot_column, K);
        // note that the result is stored at i-1 as the first
        // public key element is just a seed
        result_g.compress(&mut pk.sf_g[i], &is_pivot_column);
    }
} /* end LESS_keygen */

pub fn less_sign(sig: &mut Signature, m: &[u8], sk: &PrivateKey, pk: &PublicKey) -> usize {
    let mut g0_initial_pivot_flags = [0u8; N_pad];
    let mut is_pivot_column = [0u8; N_pad];

    /*         Private key expansion        */
    /* expand sequence of seeds for private inverse-monomial matrices */
    let mut sk_shake_state = csprng_initialize::<Shake128>(&sk.compressed_sk);

    /* Generating seed for public code G_0 */
    let mut G_0_seed = [0u8; SEED_LENGTH_BYTES];
    csprng_randombytes(&mut G_0_seed, &mut sk_shake_state);

    /* The first private key monomial is an ID matrix, no need for random
     * generation, hence NUM_KEYPAIRS-1 */
    let mut private_monomial_seeds = [[0u8; PRIVATE_KEY_SEED_LENGTH_BYTES]; NUM_KEYPAIRS - 1];
    for i in 0..(NUM_KEYPAIRS - 1) {
        csprng_randombytes(&mut private_monomial_seeds[i], &mut sk_shake_state);
    }

    // generate the salt from a TRNG
    randombytes(&mut sig.salt);

    /*         Ephemeral monomial generation        */
    let mut ephem_monomials_seed = [0u8; SEED_LENGTH_BYTES];
    csprng_randombytes(&mut ephem_monomials_seed, &mut sk_shake_state);

    /* create the prng for the "blinding" monomials for the canonical form computation */
    let mut cf_seed = [0u8; SEED_LENGTH_BYTES];
    csprng_randombytes(&mut cf_seed, &mut sk_shake_state);
    let mut cf_shake_state = csprng_initialize::<Shake128>(&cf_seed);

    let mut seed_tree = [0u8; NUM_NODES_SEED_TREE * SEED_LENGTH_BYTES];
    build_ggm(&mut seed_tree, &ephem_monomials_seed, &sig.salt);

    let mut linearized_rounds_seeds = [0u8; T*SEED_LENGTH_BYTES];
    seed_leaves(&mut linearized_rounds_seeds, &seed_tree);

    // /*         Public G_0 expansion                  */
    let mut g0_rref = Matrix::<K, N_K_pad>::rand_from_seed::<Shake128>(&G_0_seed);
    let g0_column_pos: [u16; N_K_pad] = std::array::from_fn(|i| (i + K) as u16);
    for i in 0..K {
        g0_initial_pivot_flags[i] = 0;
    }
    let full_g0 = Matrix::<K, N_pad>::from_compressed_rref::<N_K_pad>(&g0_column_pos, &g0_rref);

    let mut G0 = Matrix::<K, N_pad>::default();
    let mut hasher = Sha3_256::new();
    let mut pi_tilde: [Permutation::<K>; T] = core::array::from_fn(|_| Permutation::<K>::default());
    for i in 0..T {
        let mu_tilde = Monomial::<N>::rand_from_seed::<Shake128>(&linearized_rounds_seeds[i*SEED_LENGTH_BYTES..(i+1)*SEED_LENGTH_BYTES]);
        Matrix::monomial_mul(&mut G0, &full_g0, &mu_tilde);

        let mut permuted_pivot_flag: [u8; N_pad] = [0u8; N_pad];
        for t in 0..N {
            permuted_pivot_flag[mu_tilde.perms[t] as usize] = g0_initial_pivot_flags[t];
        }

        G0.rref::<true>(&mut is_pivot_column, &mut permuted_pivot_flag, SIGN_PIVOT_REUSE_LIMIT);
        let mut a_j = MatrixNormalized::<K_pad>::from_information_set::<K, N_pad>(&G0, &is_pivot_column);
        pi_tilde[i] = Permutation::<K>::from_information_set_composition::<N, N_pad>(&mu_tilde, &is_pivot_column);


        a_j.blind(&mut cf_shake_state);
        a_j.cf();

        for i in 0..K {
            // TODO without unsafe?
            let bytes: &[u8] = unsafe {
                std::slice::from_raw_parts(a_j[i].0.as_ptr() as *const u8, 126)
            };
            hasher.update(bytes);
        }

    }
    hasher.update(m);
    hasher.update(sig.salt);
    sig.digest = *hasher.finalize().as_array().unwrap();


    let mut fixed_weight_string = [0u8; T];
    sample_challenge(&mut fixed_weight_string, &sig.digest);

    let num_seeds_publised = ggm_path(&seed_tree, &fixed_weight_string, &mut sig.seed_storage);
    let mut emitted_monoms = 0;
    for i in 0..T {
        if fixed_weight_string[i] != 0 {
            let j = fixed_weight_string[i] as usize;
            let Q_to_multiply = Monomial::<N>::rand_from_seed::<Shake128>(&private_monomial_seeds[j]);
            let mono_action = Permutation::<K>::from_composition_action(&Q_to_multiply, &pi_tilde[i]);
            sig.cf_monom_actions[emitted_monoms] = mono_action.bytes();

            emitted_monoms += 1;
        }
    }

    num_seeds_publised
}


#[cfg(test)]
mod tests {
    #[test]
    fn keygen() {

    }
}
